core/src/main/java/org/mule/runtime/core/internal/routing/Foreach.java

```diff
@@ -129,28 +132,57 @@ private void restoreVariables(Object previousCounterVar, Object previousRootMess
-                                          } else {
-                                            // Otherwise create a new message
-                                            partEventBuilder.message(Message.builder().payload(typedValue).build());
-                                          }
-                                          return partEventBuilder.addVariable(counterVariableName, count.incrementAndGet())
-                                              .build();
-                                        })
-                                        .transform(nestedChain))
-                            .flatMapMany(identity());
+                            .concatMap(typedValue -> {
+                              Builder partEventBuilder = builder(currentEvent.get());
+                              if (typedValue.getValue() instanceof EventBuilderConfigurer) {
+                                // Support EventBuilderConfigurer currently used by Batch Module
+                                ((EventBuilderConfigurer) typedValue.getValue()).configure(partEventBuilder);
+                              } else if (typedValue.getValue() instanceof Message) {
+                                // If value is a Message then use it directly conserving attributes and properties.
+                                partEventBuilder.message((Message) typedValue.getValue());
+                              } else {
+                                // Otherwise create a new message
+                                partEventBuilder.message(Message.builder().payload(typedValue).build());
+                              }
+                              return just(partEventBuilder.addVariable(counterVariableName, count.incrementAndGet()).build())
+                                  .transform(nestedChain)
+                                  .doOnNext(result -> currentEvent.set(InternalEvent.builder(result).build()));
+                            })
+                            .takeLast(1)
+                            .map(s -> InternalEvent.builder(currentEvent.get()).message(request.getMessage()).build());
+  }
+
+  private Iterator<TypedValue<?>> splitRequest(InternalEvent request) {
+    Object payloadValue = request.getMessage().getPayload().getValue();
+    if (DEFAULT_SPIT_EXPRESSION.equals(expression) && payloadValue instanceof EventBuilderConfigurerList) {
+      // Support EventBuilderConfigurerList currently used by Batch Module
+      return Iterators.transform(((EventBuilderConfigurerList<Object>) payloadValue).eventBuilderConfigurerIterator(),
+                                 input -> TypedValue.of(input));
+    } else if (DEFAULT_SPIT_EXPRESSION.equals(expression) && payloadValue instanceof EventBuilderConfigurerIterator) {
+      // Support EventBuilderConfigurerIterator currently used by Batch Module
+      EventBuilderConfigurerIterator configurerIterator = (EventBuilderConfigurerIterator) payloadValue;
+      return new Iterator<TypedValue<?>>() {
```

Comment:
consider extracting this to its own class, in order to get more readable code and clearer stsack traces